### PR TITLE
Dev/cloud vultr

### LINKED ISSUE #53677: salt-cloud: vultr provider doesn't support "firewallgroupid" functionality
### Description of Issue

Vultr does not provide an "avail_firewallgroups" method, nor does it provide the ability to define a "FirewallGroupID" value in the VM's profile.  As such, all VMs are created with a FirewallGroupID=0 (wide open)

### Setup

https://docs.saltstack.com/en/latest/ref/clouds/all/salt.cloud.clouds.vultrpy.html

### Steps to Reproduce Issue

The feature isn't listed in the docs, nor is there an avail_* method that provides the capability, nor does the word "firewall" appear anywhere in the vultr provider code.

### Versions Report

```
# salt-master --versions
/usr/lib/python2.7/site-packages/salt/scripts.py:102: DeprecationWarning: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date.  Salt will drop support for Python 2.7 in the Sodium release or later.
Salt Version:
           Salt: 2019.2.0

Dependency Versions:
           cffi: 1.6.0
       cherrypy: 5.6.0
       dateutil: 2.8.0
      docker-py: Not Installed
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: 0.26.3
        libnacl: Not Installed
       M2Crypto: 0.31.0
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.26.4
         Python: 2.7.5 (default, Apr  9 2019, 14:30:50)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.6.1810 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-957.1.3.el7.x86_64
         system: Linux
        version: CentOS Linux 7.6.1810 Core
```

### PR DESCRIPTION
**Pull request by richardsimko**
### What does this PR do?

Improves the Vultr provider for Salt Cloud by adding options for firewallgroupid (#53677) and SSH keys.

### What issues does this PR fix or reference?
Fixes: #53677

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

File: salt/cloud/clouds/vultrpy.py [hunk 1]
```diff
@@ -132,13 +156,59 @@ def avail_scripts(conn=None):
...
+    if not kwargs:
+        kwargs = {}
+
+    if 'keyname' not in kwargs:
+        log.error('A keyname is required.')
+        return False
+
+    keys = list_keypairs(call='function')
+    keyid = keys[kwargs['keyname']]['SSHKEYID']
+    log.debug('Key ID is %s', keyid)
```
Comment (Ch3LL) [inline salt/cloud/clouds/vultrpy.py:207]: I do not have a vultr cloud instance to test this against. What information would be included here? I just want to ensure we are not logging sensitive information.

File: salt/cloud/clouds/vultrpy.py [hunk 2]
```diff
@@ -282,6 +356,21 @@ def create(vm_):
...
 
+    firewall_group_id = config.get_cloud_config_value(
+        'firewall_group_id', vm_, __opts__, search_global=False, default=None,
+    )
+
+    if firewall_group_id and str(firewall_group_id) not in avail_firewall_groups():
+        log.error('Your Vultr account does not have a firewall group with ID %s', str(firewall_group_id))
+        return False
+
+    key_list = ssh_key_ids.split(',')
```
Comment (Ch3LL) [inline salt/cloud/clouds/vultrpy.py:367]: if `ssh_key_ids` is None this will fail with an AttributeError.

**Review by Ch3LL — CHANGES_REQUESTED** [review]
Just a question and one change request. We will also need some test coverage for this change.

File: salt/cloud/clouds/vultrpy.py (hunk 1 above)
Comment (richardsimko) [reply salt/cloud/clouds/vultrpy.py:207]: This is the unique ID that Vultr sets for the key. It's used in the URLs when editing the key, such as `https://my.vultr.com/sshkeys/manage/?SSHKEYID={id}` so I don't believe it's sensitive.

File: salt/cloud/clouds/vultrpy.py (hunk 2 above)
Comment (richardsimko) [reply salt/cloud/clouds/vultrpy.py:367]: Good point! I will fix this!

**Comment by richardsimko** [comment]
Thanks for the feedback. As there are no test cases at all for the Vultr client and I don't feel comfortable taking full ownership over writing those it would be very hard to write tests exclusively covering this small change, which is why I left it out.

**Comment by Ch3LL** [comment]
Here's a test to help get you started:

```
(salt-3.7) ➜  salt git:(dev/cloud-vultr) ✗ cat tests/unit/cloud/clouds/test_vultr.py
# -*- coding: utf-8 -*-

# Import Salt Libs
from salt.cloud.clouds import vultrpy as vultr

# Import Salt Testing Libs
from tests.support.unit import TestCase
from tests.support.helpers import TstSuiteLoggingHandler


class VultrTestCase(TestCase):
    """
    Unit TestCase for salt.cloud.clouds.vultr module.
    """
    def test_show_keypair_no_keyname(self):
        """
        test salt.cloud.clouds.vultr.show_keypair
        when keyname is not in kwargs
        """
        kwargs = {}
        with TstSuiteLoggingHandler() as handler:
            assert not vultr.show_keypair(kwargs)
            assert "ERROR:A keyname is required." in handler.messages
```

Let me know if you need more direction to add test coverage for the rest of the code. Thanks :)

**Comment by Ch3LL** [comment]
Also to be clear you only need to add coverage for your change, not for the entire module.

**Comment by richardsimko** [comment]
Thanks for the guidance. I added that and another test, however most of the added logic is in the `create` function which is not very testable, I don't see any of the other test providers having test coverage for this function either, probably due to the heavy coupling to other parts and the amount of mocking which would be required. Unless we completely rewrite it that will remain quite a challenge to be tested.

**Comment by Ch3LL** [comment]
I can definitely help in writing a test for create. Can you just paste here what kind of information you would expect to find in `ssh_key_names` and `firewall_group_id`

**Comment by richardsimko** [comment]
Thanks! Sure thing, `ssh_key_names` is a comma separated list like `dev1,dev2,salt-master` each representing an SSH public key which has been uploaded to Vultr and is installed on the VM when created. `firewall_group_id` is a string representing an ID of a firewall configuration in Vultr.

**Comment by Ch3LL** [comment]
Thanks, I've added this to my backlog. should be able to get to it by next week :)

**Comment by Ch3LL** [comment]
I will need to bump this to next week, apologies.

**Comment by richardsimko** [comment]
No worries! We're not really using Vultr any more so this is more in the interest of getting it merged in case someone else uses it (Which I somewhat doubt since it's nigh on unusable without using SSH keys 😅 )

**Comment by Ch3LL** [comment]
@richardsimko i went ahead and added a couple tests. Can you review and make sure the returns and args are all correct?

**Comment by richardsimko** [comment]
@Ch3LL Sorry for the late reply, everything looks good to me! 🎉

**Review by Ch3LL — DISMISSED** [review]

**Review by s0undt3ch — APPROVED** [review]

**Comment by sagetherage** [comment]
Since this is feature can you also add a Release Notes, please?

**Review by s0undt3ch — DISMISSED** [review]

**Review by Ch3LL — APPROVED** [review]

**Pull request merged**